sdks/java/extensions/sql/zetasql/src/main/java/org/apache/beam/sdk/extensions/sql/zetasql/BeamZetaSqlCatalog.java

```diff
@@ -156,6 +159,48 @@ void addFunction(ResolvedNodes.ResolvedCreateFunctionStmt createFunctionStmt) {
             ImmutableList.of(createFunctionStmt.getSignature())));
   }
 
+  void validateJavaUdf(ResolvedNodes.ResolvedCreateFunctionStmt createFunctionStmt) {
+    for (FunctionArgumentType argumentType :
+        createFunctionStmt.getSignature().getFunctionArgumentList()) {
+      Type type = argumentType.getType();
+      if (type == null) {
+        throw new UnsupportedOperationException("UDF templated argument types are not supported.");
+      }
+      validateJavaUdfZetaSqlType(type);
+    }
+    if (createFunctionStmt.getReturnType() == null) {
+      throw new NullPointerException("UDF return type must not be null.");
+    }
+    validateJavaUdfZetaSqlType(createFunctionStmt.getReturnType());
+  }
+
+  /**
+   * Throws {@link UnsupportedOperationException} if ZetaSQL type is not supported in Java UDF.
+   * Supported types are a subset of the types supported by {@link BeamJavaUdfCalcRule}.
+   */
+  void validateJavaUdfZetaSqlType(Type type) {
+    switch (type.getKind()) {
+      case TYPE_INT64:
+      case TYPE_DOUBLE:
+      case TYPE_BOOL:
+      case TYPE_STRING:
+      case TYPE_BYTES:
+        // These types are supported.
+        break;
+      case TYPE_NUMERIC:
+      case TYPE_DATE:
+      case TYPE_TIME:
+      case TYPE_DATETIME:
+      case TYPE_TIMESTAMP:
+      case TYPE_ARRAY:
+      case TYPE_STRUCT:
+        throw new UnsupportedOperationException(
+            "ZetaSQL type not allowed in Java UDF: " + type.getKind().name());
+      default:
+        throw new UnsupportedOperationException("Unknown ZetaSQL type: " + type.getKind().name());
```

Comment:
nit: possibly just make this the same exception as above? (New types are going to be added without this list being updated, this different wording could lead to user confusion.)

Reply:
Done.